File: sudokuboard.py
import numpy as np
import time

class SudokuBoard:
    SIZE = 9

    def __init__(self):
        self.board = np.zeros((self.SIZE, self.SIZE), dtype=int)

    def _parse_board(self, lines):
        for i in range(self.SIZE):
            line = lines[i]
            for j in range(self.SIZE):
                self.board[i, j] = int(line[j])

    @staticmethod
    def read_from_file(filename):
        boards = {}
        with open(filename, 'r') as file:
            lines = file.read().splitlines()
            i = 0
            while i < len(lines):
                grid_name = lines[i]
                board_lines = lines[i+1:i+1+SudokuBoard.SIZE]
                board = SudokuBoard()
                board._parse_board(board_lines)
                boards[grid_name] = board
                i += SudokuBoard.SIZE + 1
        return boards

    def _is_number_valid(self, row, col, number):
        return self._is_row_valid(row, number) and self._is_column_valid(col, number) and self._is_box_valid(row, col, number)

    def _is_row_valid(self, row, number):
        return not number in self.board[row, :]

    def _is_column_valid(self, col, number):
        return not number in self.board[:, col]

    def _is_box_valid(self, row, col, number):
        box_row = row - row % 3
        box_col = col - col % 3
        return not number in self.board[box_row:box_row+3, box_col:box_col+3]

    def print_board(self):
        for row in self.board:
            print(' '.join(map(str, row)))
        print()

    def solve(self):
        for row in range(self.SIZE):
            for col in range(self.SIZE):
                if self.board[row, col] == 0:
                    for number in range(1, self.SIZE + 1):
                        if self._is_number_valid(row, col, number):
                            self.board[row, col] = number

                            if self.solve():
                                return True
                            else:
                                self.board[row, col] = 0
                    return False
        return True
    
    import time

if __name__ == "__main__":
    try:
        boards = SudokuBoard.read_from_file("sudoku.txt")

        start_time = time.time()

        for grid_name, board in boards.items():
            print(f"Solving {grid_name}...")

            if board.solve():
                print(f"Solved {grid_name}:")
                board.print_board()
            else:
                print(f"No solution found for {grid_name}")

        end_time = time.time()
        duration = (end_time - start_time) * 1000

        print(f"Total time taken to solve all puzzles: {duration:.2f} ms")

    except FileNotFoundError as e:
        print(f"Error: {e}")

File: SudokuBoard.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class SudokuBoard {

    private static final int SIZE = 9;

    private int[][] board;

    public SudokuBoard() {
        this.board = new int[SIZE][SIZE];
    }

    private void parseBoard(Scanner sc) {
        for (int i = 0; i < SIZE; i++) {
            String line = sc.nextLine();
            for (int j = 0; j < SIZE; j++) {
                this.board[i][j] = Character.getNumericValue(line.charAt(j));
            }
        }
    }

    public static Map<String, SudokuBoard> readFromFile(String filename) throws FileNotFoundException {
        Map<String, SudokuBoard> boards = new HashMap<>();
        try (Scanner sc = new Scanner(new File(filename))) {
            while (sc.hasNextLine()) {
                String gridName = sc.nextLine();
                SudokuBoard board = new SudokuBoard();
                board.parseBoard(sc);
                boards.put(gridName, board);
            }
        }
        return boards;
    }

    public boolean isNumberValid(int row, int col, int number) {
        return isRowValid(row, number) && isColumnValid(col, number) && isBoxValid(row, col, number);
    }

    private boolean isRowValid(int row, int number) {
        for (int i = 0; i < SIZE; i++) {
            if (board[row][i] == number) {
                return false;
            }
        }
        return true;
    }

    private boolean isColumnValid(int col, int number) {
        for (int i = 0; i < SIZE; i++) {
            if (board[i][col] == number) {
                return false;
            }
        }
        return true;
    }

    private boolean isBoxValid(int row, int col, int number) {
        int boxRow = row - row % 3;
        int boxCol = col - col % 3;

        for (int i = boxRow; i < boxRow + 3; i++) {
            for (int j = boxCol; j < boxCol + 3; j++) {
                if (board[i][j] == number) {
                    return false;
                }
            }
        }
        return true;
    }

    public void printBoard() {
        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
    }

    public boolean solve() {
        for (int row = 0; row < SIZE; row++) {
            for (int col = 0; col < SIZE; col++) {
                if (board[row][col] == 0) {
                    for (int number = 1; number <= SIZE; number++) {
                        if (isNumberValid(row, col, number)) {
                            board[row][col] = number;

                            if (solve()) {
                                return true;
                            } else {
                                board[row][col] = 0; // undo the move
                            }
                        }
                    }
                    return false; // no valid numbers can be placed in this cell
                }
            }
        }
        return true; // all cells are filled
    }

    public static void main(String[] args) {
        try {
            Map<String, SudokuBoard> boards = SudokuBoard.readFromFile("sudoku.txt");

            long startTime = System.nanoTime();

            for (Map.Entry<String, SudokuBoard> entry : boards.entrySet()) {
                System.out.println("Solving " + entry.getKey() + "...");

                if (entry.getValue().solve()) {
                    System.out.println("Solved " + entry.getKey() + ":");
                    entry.getValue().printBoard();
                } else {
                    System.out.println("No solution found for " + entry.getKey());
                }

                System.out.println();
            }

            long endTime = System.nanoTime();
            long duration = (endTime - startTime) / 1000000;  // divide by 1000000 to get milliseconds.

            System.out.println("Total time taken to solve all puzzles: " + duration + " ms");

        } catch (FileNotFoundException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}
